src/analyses/libraryDesc.ml

```diff
@@ -143,3 +143,184 @@ let of_old ?(attrs: attr list=[]) (old_accesses: Accesses.old) (classify_name):
+  | Cos (fk, exp) -> Cos (CilType.Fkind.relift fk, CilType.Exp.relift exp)
+  | Sin (fk, exp) -> Sin (CilType.Fkind.relift fk, CilType.Exp.relift exp)
+  | Tan (fk, exp) -> Tan (CilType.Fkind.relift fk, CilType.Exp.relift exp)
+
+
+  let order = function
+    | Nan _ -> 1
+    | Inf _ -> 2
+    | Isfinite _ -> 3
+    | Isinf _ -> 4
+    | Isnan _ -> 5
+    | Isnormal _ -> 6
+    | Signbit _ -> 7
+    | Isgreater _ -> 8
+    | Isgreaterequal _ -> 9
+    | Isless _ -> 10
+    | Islessequal _ -> 11
+    | Islessgreater _ -> 12
+    | Isunordered _ -> 13
+    | Ceil _ -> 14
+    | Floor _ -> 15
+    | Fabs _ -> 16
+    | Fmax _ -> 17
+    | Fmin _ -> 18
+    | Acos _ -> 19
+    | Asin _ -> 20
+    | Atan _ -> 21
+    | Atan2 _ -> 22
+    | Cos _ -> 23
+    | Sin _ -> 24
+    | Tan _ -> 25
+
+  let equal m1 m2 = (compare m1 m2) == 0
+  let hash = order
+
+  let cmp_fk_exp (fk1, e1) (fk2, e2) =
+    let r = (CilType.Fkind.compare fk1 fk2) in
+    if r <> 0 then
+      r
+    else
+      CilType.Exp.compare e1 e2
+
+  let cmp_exp_exp (e1, e1') (e2, e2') =
+    let r = (CilType.Exp.compare e1 e2) in
+    if r <> 0 then
+      r
+    else
+      CilType.Exp.compare e1' e2'
+
+  let cmp_fk_exp_exp (fk1, e1, e1') (fk2, e2, e2') =
+    let r = (CilType.Fkind.compare fk1 fk2) in
+    if r <> 0 then
+      r
+    else
+      cmp_exp_exp (e1, e1') (e2, e2')
+
+  let compare m1 m2 =
+    let r = Stdlib.compare (order m1) (order m2) in
+    if r <> 0 then
+      r
+    else
+      match m1, m2 with
+      | Nan fe1, Nan fe2 -> cmp_fk_exp fe1 fe2
+      | Inf fk1, Inf fk2 -> CilType.Fkind.compare fk1 fk2
+      | Isfinite e1, Isfinite e2 -> CilType.Exp.compare e1 e2
+      | Isinf e1, Isinf e2 -> CilType.Exp.compare e1 e2
+      | Isnan e1, Isnan e2 -> CilType.Exp.compare e1 e2
+      | Isnormal e1, Isnormal e2 -> CilType.Exp.compare e1 e2
+      | Signbit e1, Signbit e2 -> CilType.Exp.compare e1 e2
+      | Isgreater ee1, Isgreater ee2 -> cmp_exp_exp ee1 ee2
+      | Isgreaterequal ee1, Isgreaterequal ee2 -> cmp_exp_exp ee1 ee2
+      | Isless ee1, Isless ee2 -> cmp_exp_exp ee1 ee2
+      | Islessequal ee1, Islessequal ee2 -> cmp_exp_exp ee1 ee2
+      | Islessgreater ee1, Islessgreater ee2 -> cmp_exp_exp ee1 ee2
+      | Isunordered ee1, Isunordered ee2 -> cmp_exp_exp ee1 ee2
+      | Ceil fe1, Ceil fe2 -> cmp_fk_exp fe1 fe2
+      | Floor fe1, Floor fe2 -> cmp_fk_exp fe1 fe2
+      | Fabs fe1, Fabs fe2 -> cmp_fk_exp fe1 fe2
+      | Fmax fee1, Fmax fee2 -> cmp_fk_exp_exp fee1 fee2
+      | Fmin fee1, Fmin fee2 -> cmp_fk_exp_exp fee1 fee2
+      | Acos fe1, Acos fe2 -> cmp_fk_exp fe1 fe2
+      | Asin fe1, Asin fe2 -> cmp_fk_exp fe1 fe2
+      | Atan fe1, Atan fe2 -> cmp_fk_exp fe1 fe2
+      | Atan2 fee1, Atan2 fee2 -> cmp_fk_exp_exp fee1 fee2
+      | Cos fe1, Cos fe2 -> cmp_fk_exp fe1 fe2
+      | Sin fe1, Sin fe2 -> cmp_fk_exp fe1 fe2
+      | Tan fe1, Tan fe2 -> cmp_fk_exp fe1 fe2
+      | _ -> failwith "impossible"
```

Comment:
Could we just derive these here?

Reply:
It is possible but only if the type `math` in `LibraryDesc` is changed to use `CilType.Fkind.t` and `Basetype.CilExp.t` instead of `Cil.fkind` and `Cil.exp`.
If we change this, should we change the whole LibraryDesc to use these types instead of the `Cil.` types for consistency?

Reply:
`CilType.Fkind.t` and `CilType.Exp.t` are simply aliases for `Cil.fkind` and `Cil.exp` but come with all the necessary `Printable.S` functions. This is precisely to allow deriving from them. So changing to `CilType` is the way to go here.

Reply:
Ok, thanks for the clarification, I changed it